src/YoutubeAtomPlayer.tsx

```diff
@@ -94,6 +94,21 @@ const createOnStateChangeListener = (
             eventEmitters.forEach((eventEmitter) => eventEmitter('play'));
             progressEvents.hasSentPlayEvent = true;
 
+            /**
+             * Set a timeout to check progress again in the future
+             */
+            setTimeout(() => {
```

Comment:
Would you mind explaining why we'd need to repeat this check? Would it not be the case that every time the video state changes, `createOnStateChangeListener` is called? Like when the user plays play again (e.g. for the second time), `createOnStateChangeListener` is called and then it emits the events?

Reply:
Ahh I hadn't realised that `checkProgress()` contains a `setTimeout` which recursively calls `checkProgress()` on line `161` so I've removed this. 

And yeah you're right i think: `createOnStateChangeListener` is called every time something updates, I was just trying to make sure we still ran `checkProgress` if we were resuming rather than playing but its not necessary.

Reply:
Just picking up some questions raised here:

> i think: createOnStateChangeListener is called every time something updates

We only initialise the player (and it's listeners) once in the `Initialise player useEffect`.

So `createOnStateChangeListener` is actually only called once. We ensure the player is only initialised once by checking:

```js
if (!player.current)
```

Once the player is initialised `player.current` is set so we never re-initialise the player or its listeners.

`stateChangeListener` listener will then internally keep checking progress with a `setTimeout`.

Reply:
Thank you both for the detailed explanation. I'm just so interested to make sure I've understood this code correctly :) Do you think the steps below are now correct?
1- createOnStateChangeListener is called once within useEffect
2- createOnStateChangeListener returns a function: `(event) => {...}`
3- the function is passed to the player.on as a callback
4- the callback function listens to all YouTube IFrame Player API events (such as YT.PlayerState.ENDED)
5- so every time user plays the video or pauses or etc. the callback function is called. 
6- but for progress events, setTimeout is used because these events don't exist on youtube api out of the box

If it's not correct, could we have a small call at some point?

Reply:
I think - as far as I understand - that is exactly correct.

Reply:
Yep thats exactly correct @marjisound !

In case it helps some extra details as follows:

4) The `onStateChangeListener` is specifically called for 'stateChange' events. Correspondingly there are other event types (e.g. `ready`, `playbackRateChange`) that will fire their registered listeners. More info here:

https://developers.google.com/youtube/iframe_api_reference#Events

Reply:
Thanks both, very useful info ❤️